homeassistant/components/logger/helpers.py

```diff
@@ -0,0 +1,149 @@
+    for key, value in logpoints.items():
+        _set_log_level(logging.getLogger(key), value)
+    hass.bus.async_fire(EVENT_LOGGING_CHANGED)
+
+
+def _set_log_level(logger, level):
+    """Set the log level.
+
+    Any logger fetched before this integration is loaded will use old class.
+    """
+    getattr(logger, "orig_setLevel", logger.setLevel)(level)
+
+
+async def get_integration_loggers(hass, domain):
+    """Get loggers for an integration."""
+    loggers = [f"homeassistant.components.{domain}"]
+    with contextlib.suppress(IntegrationNotFound):
+        integration = await async_get_integration(hass, domain)
+        if integration.loggers:
+            loggers.extend(integration.loggers)
+    return loggers
+
+
+class LoggerSettings:
+    """Manage log settings."""
+
+    _stored_config: dict
+
+    def __init__(self, hass, yaml_config):
+        """Initialize log settings."""
+
+        self._yaml_config = yaml_config
+        self._default_level = logging.INFO
+        if DOMAIN in yaml_config:
+            self._default_level = yaml_config[DOMAIN][LOGGER_DEFAULT]
+        self._store = Store(hass, STORAGE_VERSION, STORAGE_KEY)
+
+    async def async_load(self):
+        """Load stored settings."""
+
+        def reset_persistence(settings):
+            """Reset persistence."""
+            if settings["persistence"] == "once":
+                settings["persistence"] = "none"
+
+        self._stored_config = await self._store.async_load()
+        if self._stored_config:
+            # Reset domains for which the overrides should only be applied once
+            self._stored_config = {
+                "logs": {
+                    domain: reset_persistence(settings)
+                    for domain, settings in self._stored_config["logs"].items()
+                }
+            }
+            await self.async_save()
+        else:
+            self._stored_config = {"logs": {}}
+
+    @callback
+    def _async_data_to_save(self):
+        """Generate data to be saved."""
+        return {
+            "logs": {
+                domain: settings
+                for domain, settings in self._stored_config["logs"].items()
+                if settings["persistence"] in ("once", "permanent")
+            }
+        }
+
+    @callback
+    def async_save(self):
+        """Save settings."""
+        self._store.async_delay_save(self._async_data_to_save, 15)
+
+    async def async_update(self, hass, domain, settings):
+        """Update settings."""
+        if settings["level"] == "default":
+            self._stored_config["logs"].pop(domain, None)
+            settings["level"] = self._default_level
+        else:
+            self._stored_config["logs"][domain] = settings
+        self.async_save()
+
+        if settings["type"] == "integration":
+            loggers = await get_integration_loggers(hass, domain)
+        else:
+            loggers = [domain]
+
+        combined_logs = {}
```

Comment:
Would be slightly easier to read if you did
`yaml_config = self._yaml_config` and used it in the loop below